### PR TITLE
Fix ga for mudlet

### PR DESCRIPTION
**Pull request by blongden**
#### Brief overview of PR changes/additions

Mudlet contains a isPrompt() function to detect prompts. This is powered by the presence of a GA on the line that sends it (most clients in 2021 suppress GA). This patch is to only send a GA if SGA is False and the line is a prompt. It also turns on FORCEDENDLINE for Mudlet as it only displays mud output if there's either a GA or a newline.

#### Motivation for adding to Evennia

Pretty sure it's broken without it.

### CODE REVIEW

**Review by Griatch — CHANGES_REQUESTED** [review]
Not sure about these changes being universally a good thing. It would appear to break telnet standard for the benefit of mudlet only ... Can you give some examples of what issues you see with tintin++ to make a change in the default needed?

File: evennia/server/portal/ttype.py
Comment (blongden) [reply evennia/server/portal/ttype.py:125]: FORCEDENDLINE is True by default (it may not have been once upon a time) so this code is not necessary. That's the only reason I removed it.

File: evennia/server/portal/ttype.py
Comment (blongden) [reply evennia/server/portal/ttype.py:122]: Fair point. Without a GA on every line Mudlet (latest version) doesn't render anything. It needs the end line in its absence, but also treats any line with GA as a prompt.

File: evennia/server/portal/ttype.py
Comment (Griatch) [reply evennia/server/portal/ttype.py:125]: Hm, true. There may be an argument to be made for keeping the line just to indicate tintin++ really wanting this value set, but I guess that's a bit superfluous ... so removing is fine then.

**Review by Griatch — CHANGES_REQUESTED** [review]
@blongden: So after reading your comments, let's summarize. Here are the possible states:
1. Evennia `FORCEDENDLINE + NO GA` - Mudlet renders normally.
2. Evennia `FORCEDENDLINE + GA` - Mudlet interprets input as a prompt
3. Evennia `NO FORCEDENDLINE + NO GA` - No output at all in mudlet?
4. Evennia `NO FORCEDENDLINE + GA` - prompt or no output in mudlet?

I can see the point of your change for the case of Mudlet. But if you think about it, this is a sort of corruption of the Telnet GA option, to only apply to prompts - which is _not_ part of the Telnet standard. Other clients may or may not follow this interpretation. I suspect this is an invalid (or at least very custom) implementation of the standard from the perspective of Mudlet honestly, but that's irrelevant to this discussion ... the issue is that our NOGA option mustn't selectively only apply to a certain thing - rather we must add more nuance to the option. So we seem to actually need two options here: 

1. `NOGOAHEAD` - this is like today and applies to all output. It defaults to `True`.
2. `PROMPTGOAHEAD` - This would default to `True` and overrides `NOGOAHEAD` setting for prompts only. One could also picture setting it to `None` and have the main `NOGOAHEAD` option also apply for prompts (I don't recall if the `options` support this natively though, so maybe something for later). 

Splitting like this would allow us to fine-tune the behavior of different clients. So for Mudlet, this would be `NOGOAHEAD=True`, `PROMPTGOAHEAD=True` (and `FORCEDENDLINE=True`).

So this PR would need modifying to add these changes. Since it's outside of the scope of the original PR, let me know if you are comfortable doing these changes or if you want me to do it (which may take a little while since I'm not working in master branch right now).

File: evennia/server/portal/ttype.py
Comment (blongden) [reply evennia/server/portal/ttype.py:125]: I'll have a check and see if that's captured in a test for tintin++, just in case.

Comment (blongden) [reply evennia/server/portal/ttype.py:125]: Actually, looking at the code now there's nothing out of the default settings for Tintin++ so I'm not sure we _need_ a test, it's defaults all the way. It appears that Mudlet is the only one that's special (though there's a prompt/new line bug in mud rammer that I'd like to fix at some point, too).

Do we want a test explicitly for tintin++ for this one even though FORCEDENDLINE is True by default anyway (I suspect many clients will need this or a GA)?

**Comment by blongden** [comment]
Latest push reverts everything to how it was but adds the NOPROMPTGOAHEAD flag to suppress GA on non-prompts.

File: evennia/server/portal/ttype.py
Comment (Griatch) [reply evennia/server/portal/ttype.py:125]: If it's all defaults now then no, I guess we can skip the tintin++ test.

**Review by Griatch — CHANGES_REQUESTED** [review]
Thanks for your fixes! I have made another review. Getting closer ;)

File: evennia/server/portal/telnet.py
Comment (blongden) [reply evennia/server/portal/telnet.py:340]: Good point.

File: evennia/server/portal/telnet.py
Comment (blongden) [reply evennia/server/portal/telnet.py:444]: Hm. If `NOGOAHEAD=True` we *don't* want to send a GA for prompts though, right? The client accepted that we wont send a GA so makes sending one superfluous.

`NOGOAHEAD=True` with `NOPROMPTGOAHEAD=True` is how it is now.
`NOGOAHEAD=False` (negotiated with client) with `NOPROMPTGOAHEAD=False` (set by ttype = MUDLET) is what we're targetting here.
`NOGOAHEAD=True` with `NOPROMPTGOAHEAD=False` would not be settable outside of manually fiddling with options, buy given we agreed to not send GA's, sending them seems a bit rude.

It may make more sense to just have three values for NOGOAHEAD... True and False as negotiated by the client, but then a value of "PROMPT" for 'send GA on prompt only'. Mixes bool and strings though which is a little dirty.

Comment (blongden) [reply evennia/server/portal/telnet.py:444]: All that being said, your example works for all of the above anyway...

`True`, `True` does not send GA ✅
`False`, `False` sends GA on prompt ✅
`False`, `True` sends GA on prompt ✅
`True`, `False` does not send GA ✅

Comment (Griatch) [reply evennia/server/portal/telnet.py:444]: `WILL/WONT GA` is a standardized telnet protocol handshake that we can't change with a third value. There is no information about prompts or anything there. So the `NOPROMPTGOAHEAD` must by necessity be something we set manually (or hard-code for mudlet, in this case). 

Before you wrote this: 
> Mudlet contains a isPrompt() function to detect prompts. This is powered by the presence of a GA on the line that sends it 
> [...] Without a GA on every line Mudlet (latest version) doesn't render anything. It needs the end line in its absence, but also treats any line with GA as a prompt.

From this I conclude that mudlet wants `NOGOAHEAD` for regular text, but `GA` for prompts (using `FORCEDENDLINE` to identify the end of regular text instead). At least that's how I understood it?

Comment (blongden) [reply evennia/server/portal/telnet.py:444]: Exactly. So NOGOAHEAD=False, NOPROMPTGOAHEAD=False and FORCEDENDLINE=True.

Comment (Griatch) [reply evennia/server/portal/telnet.py:444]: No. `NOGOAHEAD = True`, `NOPROMPTGOAHEAD=False` and `FORCEDENDLINE=True` based on what you said. If `NOGOAHEAD/NOPROMPTGOAHEAD` were always the same, why would we even need the latter?

Comment (blongden) [reply evennia/server/portal/telnet.py:444]: Ok. Starting from the top and in an effort to simplify...

If `NOGOAHEAD=True`, that means we suppress go aheads always. We do not send a GA under any circumstances.

If `NOGOAHEAD=False` we defer to NOPROMPTGOAHEAD to decide when to send one, thus:

* If `NOPROMPTGOAHEAD=True` we always send a GA (This what Mudlet does not like)

* If `NOPROMPTGOAHEAD=False` we only send a GA on a prompt (sendLine does not terminate in a GA)

I think we're actually saying the same thing but clashing on what NOPROMPTGOAHEAD means (for me its do we send a GA on something that is not a prompt).

Comment (Griatch) [reply evennia/server/portal/telnet.py]: I think the disparity is rather that I'm seeing it like this:

- `NOGOAHEAD=True/False` controls the GA state for the regular text output.
- `NOPROMPTGOAHEAD` controls the GA state for prompts only. This means that this can be `False/True` and still send `GA` even if `NOGOAHEAD=True`. 

I guess both approaches work, but it seems strange that `NOPROMPTGOAHEAD` would control non-prompt output when its very name indicates it's prompt-related.

Comment (blongden) [reply evennia/server/portal/telnet.py]: I do see your point but that would mean that to suppress GA on a regular (not a prompt) output, we'd need to change the negotiated NOGOAHEAD value to True for Mudlet, and set NOPROMPTGOAHEAD to False. Is that what we want?

Comment (Griatch) [reply evennia/server/portal/telnet.py]: I would imagine mudlet could send this itself, but yes, this sounds like the most logical solution, since NOGOAHEAD=True is really the most commonly expected setting.

Comment (blongden) [reply evennia/server/portal/telnet.py]: I'm sure it could but by default it wont, and expects the GA-lite experience (only GA's on prompts rather than after everything). Latest push makes it work as described above.

File: evennia/server/portal/ttype.py
Comment (blongden) [inline evennia/server/portal/ttype.py:123]: Motivation here is just in case mudlet accpets that GA is bad, at some point and just uses the lack of a terminating EOL like most other modern clients... :)

**Review by Griatch — APPROVED** [review]
Looks good now, thanks for fixing! Hopefully this makes mudlet work better out of the box now. :)

**Pull request merged**
